cartographer_grpc/framework/testing/rpc_handler_test_server.h

```diff
@@ -0,0 +1,185 @@
+        break;
+    }
+    WaitForHandlerCompletion(RpcHandlerWrapper::ON_REQUEST);
+  }
+
+  // Sends a WRITES_DONE event to the handler, waits for the handler
+  // to finish processing the READS_DONE event before returning.
+  void SendWritesDone() {
+    auto message = cartographer::common::make_unique<
+        typename RpcHandlerType::RequestType>();
+    switch (rpc_method_.method_type()) {
+      case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+        CHECK(client_writer_->WritesDone()) << "WritesDone failed.";
+        break;
+      case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      case ::grpc::internal::RpcMethod::SERVER_STREAMING:
+      case ::grpc::internal::RpcMethod::NORMAL_RPC:
+        LOG(FATAL) << "Not implemented";
+        break;
+    }
+    WaitForHandlerCompletion(RpcHandlerWrapper::ON_READS_DONE);
+  }
+
+  // Sends a FINISH event to the handler under test, waits for the
+  // handler to finish processing the event before returning.
+  void SendFinish() {
+    switch (rpc_method_.method_type()) {
+      case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+        CHECK(client_writer_->Finish()) << "Finish failed.";
+        break;
+      case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+      case ::grpc::internal::RpcMethod::SERVER_STREAMING:
+      case ::grpc::internal::RpcMethod::NORMAL_RPC:
+        LOG(FATAL) << "Not implemented";
+        break;
+    }
+    WaitForHandlerCompletion(RpcHandlerWrapper::ON_FINISH);
+  }
+
+ private:
+  using ClientWriter = ::grpc::internal::ClientWriterFactory<
+      typename RpcHandlerType::RequestType>;
+  using HandlerInstantiator =
+      std::function<std::unique_ptr<RpcHandlerInterface>(
+          Rpc *const rpc, ExecutionContext *const execution_context)>;
+
+  void WaitForHandlerCompletion(RpcHandlerWrapper::RpcHandlerEvent event) {
+    CHECK_EQ(rpc_handler_event_queue_.Pop(), event);
+  }
+
+  void InstantiateReadersAndWriters() {
+    switch (rpc_method_.method_type()) {
+      case ::grpc::internal::RpcMethod::CLIENT_STREAMING:
+        if (!response_) {
+          response_ = cartographer::common::make_unique<
+              typename RpcHandlerType::ResponseType>();
+        }
+        if (!client_writer_) {
+          client_writer_ = CreateClientWriter();
+        }
+        break;
+      case ::grpc::internal::RpcMethod::SERVER_STREAMING:
+      case ::grpc::internal::RpcMethod::NORMAL_RPC:
+      case ::grpc::internal::RpcMethod::BIDI_STREAMING:
+        LOG(FATAL) << "Not implemented";
+        break;
+    }
+  }
+
+  std::unique_ptr<ClientWriter> CreateClientWriter() {
+    return std::unique_ptr<ClientWriter>(ClientWriter::Create(
+        channel_.get(), rpc_method_, &context_, response_.get()));
+  }
+
+  RpcHandlerInfo GetRpcHandlerInfo() {
+    ::grpc::internal::RpcMethod::RpcType rpc_type =
+        RpcType<typename RpcHandlerType::IncomingType,
+                typename RpcHandlerType::OutgoingType>::value;
+    return RpcHandlerInfo{
```

Comment:
move the event publisher and handler instantiator here:
auto event_publisher = [this]....
auto handler_instantiator = [this]...

Reply:
DOne.